bin/create-scxa-analytics-schema.sh

```diff
@@ -287,12 +321,32 @@ printf "\n\nCreate update processor "
 curl -X POST -H 'Content-type:application/json' --data-binary '{
   "add-updateprocessor":
   {
-    "name": "scxa_analytics_v2_dedup"
+    "name": "scxa_analytics_v3_dedup"
     "class": "solr.processor.SignatureUpdateProcessorFactory",
     "enabled": "true",
     "signatureField": "id",
     "overwriteDupes": "true",
-    "fields": "cell_id,experiment_accession",
+    "fields": "cell_id,experiment_accession,ontology_annotation",
```

Comment:
We can get many documents with the same signature. For instance, I tried E-EHCA-2 and for the first `cell_id`, 21784_6_10, we have:
```json
[
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "age",
    "characteristic_value": [
      "6 to 12 week"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "disease",
    "characteristic_value": [
      "subcutaneous melanoma"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "facs_marker",
    "characteristic_value": [
      "CD45+ CD3e- B220- CD11b+"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "individual",
    "characteristic_value": [
      "1131"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "inferred_cell_type",
    "characteristic_value": [
      "CD11b-positive macrophage/monocyte"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "organism",
    "characteristic_value": [
      "Mus musculus"
    ],
    "ontology_annotation": [
      "http://purl.obolibrary.org/obo/NCBITaxon_10090"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "organism_part",
    "characteristic_value": [
      "lymph node"
    ],
    "ontology_annotation": [
      "http://purl.obolibrary.org/obo/UBERON_0000029"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "organism_status",
    "characteristic_value": [
      "postmortem"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "sampling_site",
    "characteristic_value": [
      "lymph node"
    ],
    "ontology_annotation": [
      "http://purl.obolibrary.org/obo/UBERON_0000029"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "sex",
    "characteristic_value": [
      "female"
    ],
    "ontology_annotation": [
      "http://purl.obolibrary.org/obo/PATO_0000383"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "characteristic_name": "strain",
    "characteristic_value": [
      "C57BL/6"
    ],
    "ontology_annotation": [
      "http://www.ebi.ac.uk/efo/EFO_0004472"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "factor_name": "sampling_site",
    "factor_value": [
      "lymph node"
    ],
    "ontology_annotation": [
      "http://purl.obolibrary.org/obo/UBERON_0000029"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "factor_name": "single_cell_identifier",
    "factor_value": [
      "21784_6#10"
    ]
  },
  {
    "experiment_accession": "E-EHCA-2",
    "cell_id": "21784_6_10",
    "factor_name": "time",
    "factor_value": [
      "8 day"
    ]
  }
]
```

This results in only five documents added to the collection.

Because documents are stripped of the `ontology_annotation` field when it’s empty, only the ones with some value in that field make it to the index. Also, of the ones that do, some have the same signature (e.g. several have the same value `"ontology_annotation": ["http://purl.obolibrary.org/obo/UBERON_0000029"])`.

This makes us lose some sample characteristics (age, disease, facs_marker, individual, inferred_cell_type, organism_part, organism_status, sampling_site) and some factors (single_cell_identifier).

Either the signature in the dedupe processor needs to be changed or we assign an ID field to the JSON docs.

Reply:
Ahh, I see. I made some bad assumptions here. I'll add an ID. Otherwise, a better signature would be `experiment_accession,cell_id,characteristic_value,factor_value`? Would that work given the value fields are multi-valued?

Reply:
I think a more correct signature would be `experiment_accession,cell_id,characteristic_name,factor_name`. I don’t know if a value could be the same for two characteristics/factors.